akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/BehaviorBuilder.scala

```diff
@@ -171,6 +172,38 @@ object BehaviorBuilder {
   def create[T]: BehaviorBuilder[T] = _empty.asInstanceOf[BehaviorBuilder[T]]
 }
 
+/**
+ * The concrete behavior
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private final class BuiltBehaviorArray[T](messageHandlers: Array[Case[T, T]], signalHandlers: Array[Case[T, Signal]])
+    extends ExtensibleBehavior[T] {
+
+  override def receive(ctx: TypedActorContext[T], msg: T): Behavior[T] = receive(msg, messageHandlers)
+
+  override def receiveSignal(ctx: TypedActorContext[T], msg: Signal): Behavior[T] = receive(msg, signalHandlers)
+
+  private def receive[M](msg: M, handlers: Array[Case[T, M]]): Behavior[T] = {
+    if (handlers.length == 0) {
+      Behaviors.unhandled[T]
+    } else {
+      var i = 0
+      while (i < handlers.length) {
+        val Case(cls, predicate, handler) = handlers(i)
+        //do the case matching
+        if ((cls.isEmpty || cls.get.isAssignableFrom(msg.getClass)) && (predicate.isEmpty || predicate.get.apply(msg)))
+          return handler(msg)
+        else
+          i += 1
+      }
```

Comment:
Make it a @tailrec loop function passing the index counter and returning early on found instead? Using `return` for loop hurts my Scala-eyes :)